### DISCUSSION TITLE
Can't get capture working

### DISCUSSION THREAD

**Discussion by magicmouse** [rust-pcap/pcap#308] [Q&A]
I can list the existing devices okay using the Device::List() function, so i know that libpcap is installed and working fine on my Mac OSX machine.

But when i try to capture some packets on the "en1" interface (my wifi port which is quite active and tcpdump shows plenty of action), it just hangs. Can someone please tell me what i am doing wrong?
I don't understand how to get past square one with this library.

let mut cap = Capture::from_device("en1").unwrap()
        .open().unwrap();

    let mut npack = 0;
    while let Ok(packet) = cap.next_packet() {
        println!("received packet! {:?}", packet);
        npack += 1;
        if npack > 100 {
            break;
        }
    }

**Comment by Stargateur** [comment]
show tcpdump output

**Comment by Wojtek242** [comment]
I don't use a mac, but in the README for MAC OS we have:

> Note: A timeout of zero may cause pcap::Capture::next to hang and never return (because it waits for the timeout to expire before returning). This can be fixed by using a non-zero timeout (as the libpcap manual recommends) and calling pcap::Capture::next in a loop.

**Reply by magicmouse** [reply]
I don't see where there is a timeout parameter in next_packet...., and i am calling next_packet() in a loop. Your code could sure use more examples of usage.

**Reply by stappersg** [reply]
_patches welcome_

**Reply by Stargateur** [reply]
@magicmouse https://docs.rs/pcap/latest/pcap/struct.Capture.html#method.timeout

**Comment by magicmouse** [comment]
Evidently you have to set immediate mode. This example works fine:

let mut cap = Capture::from_device("en1")
    .unwrap()
    .immediate_mode(true)
    .open()
    .unwrap();

    let mut npack = 0;
    loop {
        let packet = cap.next_packet();
    println!("received packet! {:?}", packet);
    npack += 1;
    if npack > 100 {
        break;
    }
    } // loop
